File: zygiskd/src/root_impl/apatch.rs
use std::process::{Command, Stdio};
use std::fs::File;
use serde::Deserialize;
use crate::constants::MIN_APATCH_VERSION;

pub enum Version {
    Supported,
    TooOld,
    Abnormal,
}
fn parse_version(output: &str) -> i32 {
    let mut version: Option<i32> = None;
    for line in output.lines() {
        if let Some(num) = line.trim().split_whitespace().last() {
            if let Ok(v) = num.parse::<i32>() {
                version = Some(v);
                break;
            }
        }
    }
    version.unwrap_or_default()
}

pub fn get_apatch() -> Option<Version> {
    let output1 = Command::new("/data/adb/apd")
        .arg("-V")
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .output()
        .ok()?;
    let stdout1 = String::from_utf8(output1.stdout).ok()?;
    let version = parse_version(&stdout1);
    const MAX_OLD_VERSION: i32 = MIN_APATCH_VERSION - 1;
    match version {
        0 => Some(Version::Abnormal),
        v if v >= MIN_APATCH_VERSION && v <= 999999 => Some(Version::Supported),
        v if v >= 1 && v <= MAX_OLD_VERSION => Some(Version::TooOld),
        _ => None,
    }
}

#[derive(Deserialize)]
#[allow(dead_code)]
struct PackageConfig {
    pkg: String,
    exclude: i32,
    allow: i32,
    uid: i32,
    to_uid: i32,
    sctx: String,
}

fn read_package_config() -> Result<Vec<PackageConfig>, std::io::Error> {
    let file = File::open("/data/adb/ap/package_config")?;
    let mut reader = csv::Reader::from_reader(file);

    let mut package_configs = Vec::new();
    for record in reader.deserialize() {
        match record {
            Ok(config) => package_configs.push(config),
            Err(error) => {
                log::warn!("Error deserializing record: {}", error);
            }
        }
    }

    Ok(package_configs)
}

pub fn uid_granted_root(uid: i32) -> bool {
    match read_package_config() {
        Ok(package_configs) => {
            package_configs
                .iter()
                .find(|config| config.uid == uid)
                .map(|config| config.allow == 1)
                .unwrap_or(false)
        }
        Err(err) => {
            log::warn!("Error reading package config: {}", err);
            return false;
        }
    }
}

pub fn uid_should_umount(uid: i32) -> bool {
    match read_package_config() {
        Ok(package_configs) => {
            package_configs
                .iter()
                .find(|config| config.uid == uid)
                .map(|config| {
                    match config.exclude {
                        1 => true,
                        _ => false,
                    }
                })
                .unwrap_or(false)
        }
        Err(err) => {
            log::warn!("Error reading package configs: {}", err);
            false
        }
    }
}

// TODO: signature
pub fn uid_is_manager(uid: i32) -> bool {
    if let Ok(s) = rustix::fs::stat("/data/user_de/0/me.bmax.apatch") {
        return s.st_uid == uid as u32;
    }
    false
}

File: zygiskd/src/root_impl/apatch.c
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <errno.h>

#include "../constants.h"
#include "../utils.h"

#include "apatch.h"

enum RootImplState apatch_get_existence(void) {
  struct stat s;
  if (stat("/data/adb/apd", &s) != 0) {
    LOGE("APATCH | Failed to stat /data/adb/apd: %s\n", strerror(errno));

    return Inexistent;
  }

  char apatch_version[32];
  char *const argv[] = { "apd", "-V", NULL };

  LOGI("APATCH | Checking for apd existence\n");
  if (!exec_command(apatch_version, sizeof(apatch_version), "/data/adb/apd", argv)) {
    LOGE("APATCH | Failed to execute apd binary: %s\n", strerror(errno));
    errno = 0;

    return Inexistent;
  }

  int version = atoi(apatch_version + strlen("apd "));
  LOGI("APATCH | apd version: %d\n", version);
  if (version == 0) return Abnormal;

  if (version >= MIN_APATCH_VERSION && version <= 999999) return Supported;
  if (version >= 1 && version <= MIN_APATCH_VERSION - 1) return TooOld;

  return Inexistent;
}

struct package_config {
  uid_t uid;
  bool root_granted;
  bool umount_needed;
};

struct packages_config {
  struct package_config *configs;
  size_t size;
};

bool _apatch_get_package_config(struct packages_config *config) {
  FILE *fp = fopen("/data/adb/ap/package_config", "r");
  if (fp == NULL) {
    LOGE("APATCH | Failed to open package_config: %s\n", strerror(errno));

    return false;
  }

  char line[256];
  /* INFO: Skip the CSV header */
  fgets(line, sizeof(line), fp);

  while (fgets(line, sizeof(line), fp) != NULL) {
    config->configs = realloc(config, (config->size + 1) * sizeof(struct package_config));
    if (config->configs == NULL) {
      LOGE("APATCH | Failed to realloc package config: %s\n", strerror(errno));

      fclose(fp);

      return false;
    }

    strtok(line, ",");

    char *exclude_str = strtok(NULL, ",");
    if (exclude_str == NULL) continue;

    char *allow_str = strtok(NULL, ",");
    if (allow_str == NULL) continue;

    char *uid_str = strtok(NULL, ",");
    if (uid_str == NULL) continue;    

    config->configs[config->size].uid = atoi(uid_str);
    config->configs[config->size].root_granted = strcmp(allow_str, "1") == 0;
    config->configs[config->size].umount_needed = strcmp(exclude_str, "1") == 0;

    config->size++;
  }

  fclose(fp);

  return true;
}

bool apatch_uid_granted_root(uid_t uid) {
  struct packages_config *config = NULL;
  if (!_apatch_get_package_config(config)) {
    free(config);

    return false;
  }

  for (size_t i = 0; i < config->size; i++) {
    if (config->configs[i].uid == uid) {
      free(config);

      return config->configs[i].root_granted;
    }
  }

  free(config);

  return false;
}

bool apatch_uid_should_umount(uid_t uid) {
  struct packages_config *config = NULL;
  if (!_apatch_get_package_config(config)) {
    free(config);

    return false;
  }

  for (size_t i = 0; i < config->size; i++) {
    if (config->configs[i].uid == uid) {
      free(config);

      return config->configs[i].umount_needed;
    }
  }

  free(config);

  return false;
}

bool apatch_uid_is_manager(uid_t uid) {
  struct stat s;
  if (stat("/data/user_de/0/me.bmax.apatch", &s) == -1) return false;

  return s.st_uid == uid;
}
